src/common/System.cpp

```diff
@@ -410,19 +405,32 @@ void GenRandomBytes(void* dest, size_t size)
 	size_t bytes_written;
 	if (nacl_secure_random(dest, size, &bytes_written) != 0 || bytes_written != size)
 		Sys::Error("nacl_secure_random failed");
-#elif defined(__linux__) && defined(HAS_GETRANDOM_SYSCALL)
-	if (syscall(SYS_getrandom, dest, size, GRND_NONBLOCK) == -1)
-		Sys::Error("Failed getrandom syscall: %s", strerror(errno));
 #elif defined(__linux__)
-	int fd = open("/dev/urandom", O_RDONLY);
-	if (fd == -1)
-		Sys::Error("Failed to open /dev/urandom: %s", strerror(errno));
-	if (read(fd, dest, size) != (ssize_t) size)
-		Sys::Error("Failed to read from /dev/urandom: %s", strerror(errno));
-	close(fd);
+	ssize_t ret = getrandom(dest, size, GRND_NONBLOCK);
+	if (ret == -1)
+	{
+		if (errno == ENOSYS)
+		{
+			Log::Warn("Failed getrandom syscall: %s", strerror(ENOSYS));
+
+			int fd = open("/dev/urandom", O_RDONLY);
+			if (fd == -1)
+				Sys::Error("Failed to open /dev/urandom: %s", strerror(errno));
+			if (read(fd, dest, size) != (ssize_t) size)
+				Sys::Error("Failed to read from /dev/urandom: %s", strerror(errno));
+			close(fd);
+		}
+		else
+		{
+			Sys::Error("Failed getrandom syscall: %s", strerror(errno));
+		}
+	}
+	else if (ret != size)
+	{
+		Sys::Error("Failed getrandom syscall: %s", strerror(EAGAIN));
```

Comment:
Why `strerror(EAGAIN)`?

Reply:
If I understand the man page correctly, when `getrandom()` returns less bytes than requested, when `GRND_NONBLOCK` is not used, it fails with `EAGAIN`. But when using `GRND_NONBLOCK`, it may still return less bytes than requested without failing, in that case we better act like if it failed with `EAGAIN`, considering the `EAGAIN` error to always be considered when `getrandom()` returns less than requested.

Reply:
OK. But I would rather see a unique message at each location so that if we have a log file we can tell which operation actually failed. So for this one you could put for example `"getrandom returned insufficient data"`. And instead of `"Failed getrandom syscall: %s", strerror(ENOSYS)`, `"getrandom syscall not supported by kernel"`.

Reply:
Done.